File: python/src/problems1_100/reverse_linked_list2.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.singly_list_node import SinglyListNode


class Problem(ABC):
    @abstractmethod
    def reverse_between(
        self,
        head: SinglyListNode | None,
        left: int,
        right: int,
    ) -> SinglyListNode | None:
        pass


class Default(Problem):
    @override
    def reverse_between(
        self,
        head: SinglyListNode | None,
        left: int,
        right: int,
    ) -> SinglyListNode | None:
        if left == right:
            return head

        root: SinglyListNode = SinglyListNode(0, head)
        center_start: SinglyListNode | None = root
        center_start_previous: SinglyListNode | None = None
        center_end: SinglyListNode | None = root
        for i in range(right):
            if i < left:
                center_start_previous = center_start
                center_start = center_start.next
            center_end = center_end.next

        excess: SinglyListNode | None = center_end.next
        center_start_previous.next = None
        center_end.next = None

        if left == 1:
            head = self.reverse(center_start)
            current = head
        else:
            current = self.traverse_until_end(head)
            current.next = self.reverse(center_start)

        current: SinglyListNode = self.traverse_until_end(current)
        current.next = excess
        return head

    def traverse_until_end(self, node: SinglyListNode | None) -> SinglyListNode | None:
        current: SinglyListNode = node
        while current.next:
            current = current.next
        return current

    def reverse(self, node: SinglyListNode) -> SinglyListNode:
        current: SinglyListNode = node
        previous: SinglyListNode | None = None
        while current:
            temp = current.next
            current.next = previous
            previous = current
            current = temp
        return previous


SOLUTIONS: list[Problem] = [Default()]

File: typescript/src/problems1_100/reverse-linked-list2.ts
import SinglyListNode from '../concepts/singly-list-node';

class Problem {
  reverseBetween(
      head: SinglyListNode | undefined,
      left: number,
      right: number,
  ): SinglyListNode | undefined {
    throw new Error(`Not implemented (${head}, ${left}, ${right}).`);
  }
}

class Default extends Problem {
  reverseBetween(
      head: SinglyListNode | undefined,
      left: number,
      right: number,
  ): SinglyListNode | undefined {
    if (left === right) {
      return head;
    }

    const root: SinglyListNode = new SinglyListNode(0, head);
    let centerStart: SinglyListNode | undefined = root;
    let centerStartPrevious: SinglyListNode | undefined = undefined;
    let centerEnd: SinglyListNode | undefined = root;
    for (let i = 0; i < right; i++) {
      if (i < left) {
        centerStartPrevious = centerStart;
        centerStart = centerStart.next;
      }
      centerEnd = centerEnd.next;
    }

    const excess: SinglyListNode = centerEnd.next;
    centerStartPrevious.next = undefined;
    centerEnd.next = undefined;

    let current: SinglyListNode | undefined;
    if (left === 1) {
      head = this.reverse(centerStart);
      current = head;
    } else {
      current = this.traverseUntilEnd(head);
      current.next = this.reverse(centerStart);
    }
    current = this.traverseUntilEnd(current);
    current.next = excess;
    return head;
  }

  traverseUntilEnd(node: SinglyListNode) {
    let current: SinglyListNode = node;
    while (current.next) {
      current = current.next;
    }
    return current;
  }

  reverse(node: SinglyListNode) {
    let current: SinglyListNode | undefined = node;
    let previous: SinglyListNode | undefined = undefined;
    while (current) {
      const temp: SinglyListNode = current.next;
      current.next = previous;
      previous = current;
      current = temp;
    }
    return previous;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
